File: src-wails/hostname.go
package main

import (
	"net"
	"strings"
	"unicode"
)

const hostnameSuffix = "devportal.localhost"

func fullHostname(label string) string {
	if label == "" {
		return ""
	}
	return label + "." + hostnameSuffix
}

func publicAppURL(entry AppEntry, port *uint16, status AppStatus, gatewayPort uint16, bindHost string) *string {
	if entry.Hostname != "" {
		u := "http://" + fullHostname(entry.Hostname)
		if gatewayPort != 0 && gatewayPort != 80 {
			u += ":" + itoaPort(gatewayPort)
		}
		return &u
	}
	switch status {
	case AppStatusIdle, AppStatusStarting, AppStatusRunning:
		if port != nil {
			u := loopbackURL(bindHost, *port)
			return &u
		}
	}
	return nil
}

func hostLabelFromRequest(host string) string {
	h := strings.ToLower(strings.TrimSpace(host))
	if h == "" {
		return ""
	}
	if parsed, _, err := net.SplitHostPort(h); err == nil {
		h = parsed
	}
	h = strings.TrimSuffix(h, ".")
	suffix := "." + hostnameSuffix
	if !strings.HasSuffix(h, suffix) {
		return ""
	}
	label := strings.TrimSuffix(h, suffix)
	if label == "" || strings.Contains(label, ".") {
		return ""
	}
	return label
}

func normalizeHostname(input string) (string, error) {
	s := strings.ToLower(strings.TrimSpace(input))
	s = strings.TrimPrefix(s, "http://")
	s = strings.TrimPrefix(s, "https://")
	s = strings.TrimSuffix(s, "/")
	if host, _, err := net.SplitHostPort(s); err == nil {
		s = host
	}
	s = strings.TrimSuffix(s, "."+hostnameSuffix)
	s = strings.Trim(s, ".")
	if s == "" {
		return "", nil
	}
	if strings.Contains(s, ".") {
		return "", errString("ドメインは " + hostnameSuffix + " の直前の1ラベルだけ指定してください")
	}
	if !validHostnameLabel(s) {
		return "", errString("ドメインが不正です: " + s + "（英小文字・数字・ハイフン、先頭末尾はハイフン不可）")
	}
	return s, nil
}

func validHostnameLabel(s string) bool {
	if s == "" || len(s) > 63 {
		return false
	}
	if s[0] == '-' || s[len(s)-1] == '-' {
		return false
	}
	for i := 0; i < len(s); i++ {
		c := s[i]
		if c >= 'a' && c <= 'z' || c >= '0' && c <= '9' || c == '-' {
			continue
		}
		return false
	}
	return true
}

func slugHostname(s string) string {
	s = strings.ToLower(strings.TrimSpace(s))
	var b strings.Builder
	lastDash := false
	for _, r := range s {
		switch {
		case r >= 'a' && r <= 'z' || r >= '0' && r <= '9':
			b.WriteRune(r)
			lastDash = false
		case unicode.IsSpace(r) || r == '_' || r == '.' || r == '/':
			if b.Len() > 0 && !lastDash {
				b.WriteByte('-')
				lastDash = true
			}
		default:
			if b.Len() > 0 && !lastDash {
				b.WriteByte('-')
				lastDash = true
			}
		}
	}
	out := strings.Trim(b.String(), "-")
	if len(out) > 63 {
		out = strings.Trim(out[:63], "-")
	}
	if out == "" {
		return "app"
	}
	return out
}

func uniqueHostname(apps []AppEntry, id, preferred string) string {
	if preferred == "" {
		return ""
	}
	candidate := preferred
	n := 2
	for {
		taken := false
		for _, app := range apps {
			if app.ID != id && app.Hostname == candidate {
				taken = true
				break
			}
		}
		if !taken {
			return candidate
		}
		suffix := "-" + itoaPort(uint16(n))
		base := preferred
		if len(base)+len(suffix) > 63 {
			base = strings.Trim(base[:63-len(suffix)], "-")
		}
		candidate = base + suffix
		n++
	}
}

File: src/lib/hostname.ts
const HOSTNAME_SUFFIX = "devportal.localhost";

export function hostnameFieldError(input: string): string | null {
  const s = normalizeHostnameInput(input);
  if (s === "") return null;
  if (s.includes(".")) {
    return `ドメインは ${HOSTNAME_SUFFIX} の直前の1ラベルだけ指定してください`;
  }
  if (!validHostnameLabel(s)) {
    return `ドメインが不正です: ${s}（英小文字・数字・ハイフン、先頭末尾はハイフン不可）`;
  }
  return null;
}

export function isHostnameError(message: string): boolean {
  return message.includes("ドメイン");
}

function normalizeHostnameInput(input: string): string {
  let s = input.trim().toLowerCase();
  s = s.replace(/^https?:\/\//, "");
  s = s.replace(/\/$/, "");
  const colon = s.lastIndexOf(":");
  if (colon > 0 && /^\d+$/.test(s.slice(colon + 1))) {
    s = s.slice(0, colon);
  }
  if (s.endsWith("." + HOSTNAME_SUFFIX)) {
    s = s.slice(0, -(HOSTNAME_SUFFIX.length + 1));
  }
  return s.replace(/^\.+|\.+$/g, "");
}

function validHostnameLabel(s: string): boolean {
  if (s.length === 0 || s.length > 63) return false;
  if (s.startsWith("-") || s.endsWith("-")) return false;
  return /^[a-z0-9-]+$/.test(s);
}

export function slugHostname(input: string): string {
  const source = input.trim().toLowerCase();
  let out = "";
  let lastDash = false;
  for (const ch of source) {
    const alnum = (ch >= "a" && ch <= "z") || (ch >= "0" && ch <= "9");
    if (alnum) {
      out += ch;
      lastDash = false;
      continue;
    }
    if (out.length > 0 && !lastDash) {
      out += "-";
      lastDash = true;
    }
  }
  let result = out.replace(/^-+|-+$/g, "");
  if (result.length > 63) {
    result = result.slice(0, 63).replace(/-+$/g, "");
  }
  return result === "" ? "app" : result;
}

export function uniqueHostname(
  taken: readonly string[],
  preferred: string,
): string {
  if (preferred === "") return "";
  const used = new Set(taken);
  let candidate = preferred;
  let n = 2;
  while (used.has(candidate)) {
    const suffix = `-${n}`;
    let base = preferred;
    if (base.length + suffix.length > 63) {
      base = base.slice(0, 63 - suffix.length).replace(/-+$/g, "");
    }
    candidate = `${base}${suffix}`;
    n += 1;
  }
  return candidate;
}
